src/Mvc/Mvc.Core/src/ModelBinding/Binders/CollectionModelBinder.cs

```diff
@@ -140,7 +140,7 @@ public virtual async Task BindModelAsync(ModelBindingContext bindingContext)
                     AddErrorIfBindingRequired(bindingContext);
                 }
 
-                bindingContext.Result = ModelBindingResult.Success(model);
+                bindingContext.Result = model == null ? ModelBindingResult.Failed() : ModelBindingResult.Success(model);
```

Comment:
Am I correct that before .NET 7, given a`Get([FromQuery] Dictionary<string, string> test = null)` action and no "test"entries in the query string, you would get an empty collection instead of null? And in .NET 7 before this fix, you get a NullReferenceException before the action gets invoked?

If so, why didn't `CollectionModelBinder_DoesNotCreateEmptyCollection_IfModelHasDefaultValue` catch this? Do we need more end-to-end testing here? Are we going to backport this fix?

Reply:
> Am I correct that before .NET 7, given aGet([FromQuery] Dictionary<string, string> test = null) action and no "test"entries in the query string, you would get an empty collection instead of null? And in .NET 7 before this fix, you get a NullReferenceException before the action gets invoked?

You are right.

> If so, why didn't CollectionModelBinder_DoesNotCreateEmptyCollection_IfModelHasDefaultValue catch this? Do we need more end-to-end testing here?

the problem is the `NullReference` happens in the `DictionaryBinder` that call the base.BindModelAsync (`CollectionModelBinder.BindModelAsync`) and was expecting a non-null IDictionary whe the `IsModelSet` is `true`.

This new test `DictionaryModelBinder_DoesNotCreateEmptyCollection_IfIsTopLevelObjectAndHasDefaultValue` would have caught this.

>  Are we going to backport this fix?

I believe we should.